src/dbus/xdg_deskop_portal_impl_access.rs

```diff
@@ -0,0 +1,150 @@
+use std::ops::Not;
+
+use zbus::dbus_interface;
+
+use super::Start;
+
+pub struct AccessPortalImpl {
+    to_niri: calloop::channel::Sender<AccessDialogRequest>,
+}
+
+#[dbus_interface(name = "org.freedesktop.impl.portal.Access")]
+impl AccessPortalImpl {
+    /// AccessDialog method
+    #[allow(clippy::too_many_arguments)]
+    async fn access_dialog(
+        &self,
+        _handle: zbus::zvariant::ObjectPath<'_>,
+        app_id: &str,
+        parent_window: &str,
+        title: &str,
+        subtitle: &str,
+        body: &str,
+        options: std::collections::HashMap<&str, zbus::zvariant::Value<'_>>,
+    ) -> zbus::fdo::Result<(
+        u32,
+        std::collections::HashMap<String, zbus::zvariant::OwnedValue>,
+    )> {
+        let (response_channel_sender, response_channel_receiver) = async_channel::bounded(1);
+
+        let options = AccessDialogOptions::from_options(options);
+        let request = AccessDialogRequest {
+            app_id: app_id.to_string(),
+            parent_window: parent_window
+                .is_empty()
+                .not()
+                .then(|| parent_window.to_string()),
+            title: title.to_string(),
+            subtitle: subtitle.to_string(),
+            body: body.is_empty().not().then(|| body.to_string()),
+            options,
+
+            response_channel_sender,
+        };
+
+        if let Err(err) = self.to_niri.send(request) {
+            tracing::warn!(?err, "failed to send access dialog request");
+            return Ok((2, Default::default()));
```

Comment:
Niri-side errors should return an error from the portal. Something like:
```
return Err(fdo::Error::Failed(format!(
    "error creating session object: {err:?}"
)));
```

This is especially important for the access portal because an error return won't store the permission, whereas a successful return with nonzero value will store the denied permission: https://github.com/flatpak/xdg-desktop-portal/blob/bfb1c08718f91ae91b468ce309366120f5b8a162/src/camera.c#L141

Reply:
Ah, okay, I interpreted the docs at https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.Request.html#org-freedesktop-portal-request-response as this should return 2 in this case and will not store a decision.

changed